### PR TITLE
Improve IPAddressFamily checks

### PR DESCRIPTION
**Pull request by grahamwoodward**
Address the insufficient IPAddressFamily checking #18526 

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by levitte** [comment]
This PR only adds an empty line.  I assume that was a mistake

**Comment by grahamwoodward** [comment]
> This PR only adds an empty line. I assume that was a mistake

yeah it's by no means finished yet

**Comment by t8m** [comment]
Please do not add merge commits. Instead always use git rebase to rebase on top of the current master branch.

**Comment by grahamwoodward** [comment]
> Please do not add merge commits. Instead always use git rebase to rebase on top of the current master branch.

Yeah sorry about that, that was a mistake

**Comment by grahamwoodward** [comment]
I'm trying to write a test case for this and so I'm creating 2 IPAddressFamily objects, but with different key/keylengths to show how we're not capping the Len but I'm not sure how I can get the existing IPAddressFamily_cmp() function to run (it's static within v3_addr.c so I can't call it from my test case)

**Comment by grahamwoodward** [comment]
I want to write a test in test/v3ext.c but I'm unsure how I can get access to the static IPAddressFamily_cmp() function without making it public

**Comment by grahamwoodward** [comment]
@mattcaswell do I need to do anything to get this ready for review? also do you know how I could expose the IPAddressFamily_cmp to test it?

**Comment by botovq** [comment]
There is no need to expose it only for a test. Once you have a canonized `IPAddrBlocks`, say `addr`, you can use something like

```cmp_func = sk_IPAddressFamily_set_cmp_func(addr, NULL);```

to retrieve a function pointer to `IPAddressFamily_cmp()` (it should then be restored on `addr` if it is used otherwise).

However, I don't think your capping the length to 3 addresses the problem at hand. If the AFI/SAFI is too long, this will still be quietly ignored and if it is too short, that will also still be accepted.

Since a comparison function can't fail, invalid `IPAddressFamily` (i.e. those with AFI/SAFI of length other than 2 or 3) need to be caught before `IPAddressFamily_cmp()` is called on them.

**Comment by wan7teen-ping** [comment]
Hi I will help you with everything you need but you have to subscribe to
the youtube channel and I will fulfill all your requests

https://youtu.be/aIaqr4eq18g

Pada Isn, 26 Sep 2022, 2:21 AM Graham Woodward ***@***.***>
menulis:

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by grahamwoodward** [comment]
@mattcaswell are you OK with this PR? thanks

**Comment by FdaSilvaYY** [comment]
No merge commit please.
Just 'git rebase' if there is any conflict.

**Comment by grahamwoodward** [comment]
> No merge commit please.
> Just 'git rebase' if there is any conflict.

yeah I didn't mean to do this sorry.

**Comment by grahamwoodward** [comment]
having trouble writing the tests for this, adding to the existing test/v3ext.c tests...I'm trying to craft my own IPAddressFamily ptr with a > 3 keyless, which is easy enough and then push that onto an IPAddressBlocks object but the push fails, which I don't understand...or rather because I'm pushing, I'm not cleaning up correctly?

**Comment by mattcaswell** [comment]
if (!sk_IPAddressFamily_push(addr, f1))
        goto end;

This pushes `f1` into the `addr` stack.

    IPAddressFamily_free(f1);
    sk_IPAddressFamily_pop_free(addr, IPAddressFamily_free);

The first line frees `f1`. The second line calls `IPAddressFamily_free` for all elements in the stack, and then frees the stack itself. But since `f1` is in the stack and we already freed it, that means `IPAddressFamily_free` will be called again for `f1` resulting in a double free.

**Comment by grahamwoodward** [comment]
> ```
> if (!sk_IPAddressFamily_push(addr, f1))
>     goto end;
> ```
> 
> 
>      
>   
> 
> This pushes `f1` into the `addr` stack.
> 
> 
> ```
> IPAddressFamily_free(f1);
> sk_IPAddressFamily_pop_free(addr, IPAddressFamily_free);
> ```
> 
> 
>      
>   
> 
> The first line frees `f1`. The second line calls `IPAddressFamily_free` for all elements in the stack, and then frees the stack itself. But since `f1` is in the stack and we already freed it, that means `IPAddressFamily_free` will be called again for `f1` resulting in a double free.

I've tried removing the IPAddressFamily_free(f1); and it still fails

**Comment by mattcaswell** [comment]
I'd suggest running it in a debugger to figure out what's going wrong.

**Comment by slontis** [comment]
If you have a look at test/run_tests.pl you can see what environment variables normally get set when you run the tests.. i.e. If your current working dir is the root.
````
export LD_LIBRARY_PATH=.
export OPENSSL_CONF_INCLUDE=test
export OPENSSL_MODULES=providers
export CTLOG_FILE=test/ct/log_list.cnf
````
and then just run  gdb 
on ./test/testname

If the test recipe has additional args then you need to supply these to gdb also.,..


There are 2 types of tests: internal and external.. Search for 'internal' in test/build.info for examples..

There are quite a few functions that are exposed internally... These use the name ossl_

e.g see ossl_a2i_ipadd in v3_utl.c.. This is included in include/crypto/x509.h..

To access these includes the test/build.info has a INCLUDE[some_test]=... which specifies the include paths.

**Comment by grahamwoodward** [comment]
I don't understand the IPAddressFamily_cmp() and why we say if the return is >= 0, then it's a failure

The memcmp() returns 0 if the strings are identical and in my test case + extra tracing I can see the two strings are identical and so cmp is 0, the return statement:
```
return cmp ? cmp : a->length - b->length` 
```
results in 0 being returned

but the addr_is_canonical() sees that as a failure?

```
if (IPAddressFamily_cmp(&a, &b) >=0)
  return 0;

```

**Comment by botovq** [comment]
RFC 3779, section 2.2.3.3: There MUST be only one IPAddressFamily
SEQUENCE per unique combination of AFI and SAFI.

**Comment by grahamwoodward** [comment]
I don't think there's any docs to update for this PR

File: crypto/x509/v3_addr.c [hunk 1]
```diff
@@ -692,11 +692,20 @@ static int IPAddressFamily_cmp(const IPAddressFamily *const *a_,
     const ASN1_OCTET_STRING *a = (*a_)->addressFamily;
     const ASN1_OCTET_STRING *b = (*b_)->addressFamily;
     int len = ((a->length <= b->length) ? a->length : b->length);
+
```
Comment (t8m) [inline crypto/x509/v3_addr.c:695]: Nit: Please drop this empty line

Comment (t8m) [reply crypto/x509/v3_addr.c:695]: Ouch! You've removed the other empty line. You should've removed the empty line **above** int cmp declaration not below.

**Comment by t8m** [comment]
@beldmit still OK?

**Comment by beldmit** [comment]
Yes, still OK.

**Comment by grahamwoodward** [comment]
are those two erroring checks related to my change?

**Comment by t8m** [comment]
If you see just the 4 commits in the checked out branch then `git push --force` should do. Otherwise `git reset --hard HEAD~` and `git push --force`.

**Comment by grahamwoodward** [comment]
> But you can leave this as is - we will drop the merge commit when merging. We will need to squash these 4 commits into one or two anyway.

OK thanks for the help. Do I need to do anything else with this or just wait for it to get merged?

**Comment by t8m** [comment]
Just wait. :grin:

File: test/v3ext.c [hunk 2]
```diff
@@ -226,6 +226,92 @@ static int test_addr_ranges(void)
...
+    if (f1->ipAddressChoice == NULL &&
+        (f1->ipAddressChoice = IPAddressChoice_new()) == NULL)
+        goto end;
+    if (f1->addressFamily == NULL &&
+        (f1->addressFamily = ASN1_OCTET_STRING_new()) == NULL)
+        goto end;
+    if (!ASN1_OCTET_STRING_set(f1->addressFamily, key, keylen))
+        goto end;
+
+    /* Mark this as inheritance so we skip some of the is_cananize checks */
```
Comment (FdaSilvaYY) [inline test/v3ext.c:298]: Typo : is_cananize ?

**Comment by grahamwoodward** [comment]
typo fixed, can someone with the right powers add the ready to merge?

**Review by t8m — APPROVED** [review]
Still OK

**Comment by t8m** [comment]
@beldmit please reconfirm

**Comment by t8m** [comment]
Merging as this was trivial change so no need to restart the 24h period.

**Comment by t8m** [comment]
Merged to master, 3.1, and 3.0 branches. Thank you for your contribution.

**Pull request closed without merging**